Scripts/Modules/SnappingModule/SnappingModule.cs

```diff
@@ -0,0 +1,709 @@
+				var directionVector = GetDirection(i);
+				var rotationOffset = GetRotation(i);
+				var ray = new Ray(targetPosition + targetRotation * boundsCenter, targetRotation * directionVector);
+
+				var boundsExtents = bounds.extents;
+				var projectedExtents = Vector3.Project(boundsExtents, directionVector);
+				var raycastDistance = projectedExtents.magnitude * k_DirectSurfaceSearchScale;
+				var offset = -boundsCenter;
+				if (i > 2)
+					offset -= projectedExtents;
+				else
+					offset += projectedExtents;
+
+				if (PerformSurfaceSnapping(ray, ref position, ref rotation, targetPosition, state, offset, targetRotation, rotationOffset, upVector, m_IgnoreList, breakDistance, raycastDistance))
+					return true;
+			}
+
+			if (TryBreakSurfaceSnapping(ref position, ref rotation, targetPosition, targetRotation, state, breakDistance))
+				return true;
+
+			return false;
+		}
+
+		static Vector3 GetDirection(int i)
+		{
+			switch (i)
+			{
+				default:
+					return Vector3.down;
+				case 1:
+					return Vector3.left;
+				case 2:
+					return Vector3.back;
+				case 3:
+					return Vector3.right;
+				case 4:
+					return Vector3.forward;
+				case 5:
+					return Vector3.up;
+			}
+		}
+
+		static Quaternion GetRotation(int i)
+		{
+			switch (i)
+			{
+				default:
+					return Quaternion.AngleAxis(90, Vector3.right);
+				case 1:
+					return Quaternion.AngleAxis(90, Vector3.down);
+				case 2:
+					return Quaternion.identity;
+				case 3:
+					return Quaternion.AngleAxis(90, Vector3.up);
+				case 4:
+					return Quaternion.AngleAxis(180, Vector3.up);
+				case 5:
+					return Quaternion.AngleAxis(90, Vector3.left);
+			}
+		}
+
+		static Vector3 GetUpVector(int i, Quaternion rotation)
+		{
+			switch (i)
+			{
+				default:
+					return rotation * Vector3.back;
+				case 1:
+					return rotation * Vector3.up;
+				case 2:
+					return rotation * Vector3.up;
+				case 3:
+					return rotation * Vector3.up;
+				case 4:
+					return rotation * Vector3.up;
+				case 5:
+					return rotation * Vector3.forward;
+			}
+		}
+
+		static Color GetColor(int i)
+		{
+			switch (i)
+			{
+				default:
+					return Color.yellow;
+				case 1:
+					return Color.cyan;
+				case 2:
+					return Color.blue;
+				case 3:
+					return Color.red;
+				case 4:
+					return Color.magenta;
+				case 5:
+					return Color.green;
+			}
+		}
+
+		static bool TryBreakSurfaceSnapping(ref Vector3 position, ref Quaternion rotation, Vector3 targetPosition, Quaternion targetRotation, SnappingState state, float breakDistance)
+		{
+			if (state.surfaceSnapping)
+			{
+				if (Vector3.Distance(state.snappedPosition, targetPosition) > breakDistance)
+				{
+					position = targetPosition;
+					rotation = targetRotation;
+					state.surfaceSnapping = false;
+				}
+
+				return true;
+			}
+			return false;
+		}
+
+		static void SetupIgnoreList(List<GameObject> ignoreList, Renderer[] playerHeadObjects, GameObject go)
```

Comment:
I don't like that we are referring directly to "player head" in our snapping module. Can you come up with an uncoupled way to do this?